src/condition/sqli_detector.cpp

```diff
@@ -0,0 +1,559 @@
+{
+    return string_iequals(str, "ASC") || string_iequals(str, "DESC");
+}
+
+bool is_limit_or_offset(std::string_view str)
+{
+    return string_iequals(str, "LIMIT") || string_iequals(str, "OFFSET");
+}
+
+} // namespace
+
+std::string strip_literals(std::string_view statement, std::span<sql_token> tokens)
+{
+    std::string stripped;
+    stripped.reserve(statement.size());
+
+    auto current_token = tokens.begin();
+    if (current_token == tokens.end()) {
+        // Shouldn't happen
+        return {};
+    }
+
+    for (std::size_t i = 0; i < statement.size();) {
+        auto token_begin = current_token->index;
+        if (i < token_begin) {
+            // Characters unrelated to the current token
+            stripped += statement.substr(i, token_begin - i);
+            i = token_begin;
+            continue;
+        }
+
+        auto token_end = token_begin + current_token->str.size();
+        if (i >= token_begin && i < token_end) {
+            // We're within the current token, so let's process it
+            if (is_literal(current_token->type)) {
+                // Since this is a literal, we replace it and skip to the next
+                // character and token
+                stripped += "?";
+
+            } else {
+                // Not a literal, copy the whole thing and move on
+                stripped += statement.substr(token_begin, current_token->str.size());
+            }
+
+            i = token_end;
+        }
+
+        // At this stage we could be past the current token or we have just
+        // copied it, in either case we move forward and reevaluate
+        if (++current_token == tokens.end()) {
+            // Since there are no more tokens, we copy any remaining characters
+            if (token_end < statement.size()) {
+                stripped += statement.substr(i);
+            }
+
+            break;
+        }
+    }
+
+    return stripped;
+}
+
+bool is_query_comment(std::span<sql_token> tokens)
+{
+    /* We want to consider as malicious user-injected comments, even if they
+     * are below our 3 tokens limit.
+     *
+     * Some examples of comments:
+     *
+     *     SELECT * FROM t WHERE id =123 # AND pwd = HASH(pwd)
+     *                               \----/
+     *                           injected string
+     *     SELECT * FROM t WHERE id =1-- AND pwd = HASH(pwd)
+     *                              \---/
+     *                            injected string
+     */
+    for (std::size_t i = 1; i < tokens.size(); ++i) {
+        if (tokens[i].type == sql_token_type::eol_comment) {
+            return true;
+        }
+    }
+
+    // TODO: the slash star comment that could be injected from 2 distinct
+    // places:
+    //
+    //     SELECT * FROM t WHERE id =X AND pwd = HASH(pwd) AND country=Y
+    //
+    // Could be injected in:
+    //
+    //   SELECT * FROM t WHERE id =1 /* AND pwd = HASH(pwd) AND country=*/
+    //
+    // So a part of the query is escaped.
```

Comment:
This was a comment I copied over from the original code, indeed this would be considered an untermianted comment (albeit not eol), I haven't added a case for it, hence why I left the `TODO`.